sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubListener.PartitionProcessor.cs

```diff
@@ -218,13 +219,29 @@ public async Task ProcessEventsAsync(EventProcessorHostPartition context, IEnume
                 // so that is the responsibility of the user's function currently. E.g.
                 // the function should have try/catch handling around all event processing
                 // code, and capture/log/persist failed events, since they won't be retried.
-                if (eventToCheckpoint != null)
+                if (eventToCheckpoint == null)
                 {
-                    await CheckpointAsync(eventToCheckpoint, context).ConfigureAwait(false);
+                    return;
                 }
+
+                // If we are shutting down (or have lost ownership) and the function invocation failed, we should not checkpoint
+                // as there is a possibility the failure caused the user's function to not execute. We use the token rather than
+                // the _stoppedReason because the _stoppedReason is not set until after the partition processor has stopped.
+                if (linkedCts.IsCancellationRequested && !result.Succeeded)
+                {
+                    return;
+                }
+
+                // If ownership of the partition is lost, we should not checkpoint as the event may need to be processed again.
+                if (_stoppedReason is ProcessingStoppedReason.OwnershipLost)
```

Comment:
My assumption was that if we are shutting down, it is safe to checkpoint as the partition processor still has ownership over the partition being processed. If the user threads the CT through any processing they are doing, and there function throws, the execution will have failed and we will not checkpoint.

Reply:
My assumption was that if we are shutting down, it is safe to checkpoint as the partition processor still has ownership over the partition being processed. If the user threads the CT through any processing they are doing, and there function throws, the execution will have failed and we will not checkpoint.

Reply:
That isn't correct.  Ownership changes are not coordinated between processors.  The new owner will take over before the old owner is aware - either via load balancing or being denied the next read attempt.  We can't assume that developers honoring the cancellation token will throw.  It is quite common to see the pattern:

```csharp
if (cancellationToken.IsCancellationRequested)
{
    return;
}
```

Worst case, we emit a checkpoint and the new owner hasn't initialized yet.  The new owner reads our checkpoint and the batch that we just skipped because cancellation was requested is now lost.